### PR TITLE
Support adding a new column with `id` and `name` column mapping mode in Delta Lake

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Support adding a new column with `id` and `name` column mapping mode in Delta Lake

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Support adding a new column with `id` and `name` column mapping mode. ({issue}`17236`)
```

### CODE REVIEW

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java [hunk 1]
```diff
@@ -290,6 +305,54 @@ public void testColumnMappingModeAddColumn(String mode)
...
+                            row("a_row", "row(x integer)", "", ""));
+
+            // TODO: Insert on Trino once the connector supports id column mapping mode
+            onDelta().executeQuery("INSERT INTO default." + tableName + " VALUES (2, array(3), map('key', 4), named_struct('x', 5))");
+
+            List<Row> expectedRows = ImmutableList.of(row(1, null, null, null), row(2, 3, 4, 5));
+            assertThat(onDelta().executeQuery("SELECT a_number, a_array[0], a_map['key'], a_row.x FROM default." + tableName)).containsOnly(expectedRows);
+            assertThat(onTrino().executeQuery("SELECT a_number, a_array[1], a_map['key'], a_row.x FROM delta.default." + tableName)).containsOnly(expectedRows);
+
+            // 8 comes from 1 (a_number) + 2 (column & element of a_array) + 3 (column, key & value of a_map) + 2 (column & field of a_row)
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:347]: We could list here the output of `parquet-tools schema ...` for an output file to add more clarity what the column id signifies.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 2]
```diff
@@ -125,6 +128,12 @@ public static ColumnMappingMode getColumnMappingMode(MetadataEntry metadata)
         return Enums.getIfPresent(ColumnMappingMode.class, columnMappingMode.toUpperCase(ENGLISH)).or(ColumnMappingMode.UNKNOWN);
     }
 
+    public static int getMaxColumnId(MetadataEntry metadata)
+    {
+        String maxColumnId = metadata.getConfiguration().get(MAX_COLUMN_ID_CONFIGURATION_KEY);
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:133]: null check?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java [hunk 3]
```diff
@@ -290,6 +305,54 @@ public void testColumnMappingModeAddColumn(String mode)
...
+                " 'delta.columnMapping.mode' = '" + mode + "'" +
+                ")");
+
+        try {
+            // TODO: Insert on Trino once the connector supports id column mapping mode
+            onDelta().executeQuery("INSERT INTO default." + tableName + " VALUES 1");
+            assertThat(onDelta().executeQuery("SELECT * FROM default." + tableName)).containsOnly(row(1));
+            assertThat(onTrino().executeQuery("SELECT * FROM delta.default." + tableName)).containsOnly(row(1));
+
+            onTrino().executeQuery("ALTER TABLE delta.default." + tableName + " ADD COLUMN a_array array(integer)");
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:329]: Let's add a bit more nesting (array of struct, map<string, struct>, struct containing structs) to make sure that generating the new delta lake schema is accurate

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 4]
```diff
@@ -142,21 +151,27 @@ public static List<DeltaLakeColumnHandle> extractPartitionColumns(List<DeltaLake
     }
 
     public static String serializeSchemaAsJson(
+            ColumnMappingMode columnMappingMode,
             List<DeltaLakeColumnHandle> columns,
+            AtomicInteger newFieldId,
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:156]:
```suggestion
OptionalInt newFieldId,
```

Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:156]: The logic depends on `AtomicInteger` to propagate the max id. I won't address this comment.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java (hunk 3 above)
Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:329]: Why adding more nested fields ensures the schema is accurate?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java (hunk 1 above)
Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:347]: That's overkill. I won't do that in this PR.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java (hunk 3 above)
Comment (findinpath) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:329]: Fair question. This part has already been covered by other tests.
I wanted to make sure that the column mapping does indeed get added also on nested field types.

Comment (findinpath) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:329]:
```
spark-sql> create table tiny.test2(c1 integer) using delta location 's3://trino-ci-test/test2' tblproperties ('delta.columnMapping.mode'='id');

spark-sql> insert into tiny.test2 values 1,2;
Time taken: 1.649 seconds
spark-sql> select * from tiny.test2;
1	NULL	NULL
2	NULL	NULL

```

```
trino> alter table delta.tiny.test2 add column c2 varchar;
ADD COLUMN
trino> alter table delta.tiny.test2 add column c3 array(row(leaf1 integer, leaf2 integer));
ADD COLUMN
trino> select * from delta.tiny.test2;
 c1 |  c2  |  c3  
----+------+------
  2 | NULL | NULL 
  1 | NULL | NULL 
(2 rows)
```

```
spark-sql> insert into tiny.test2 values (3, 33, array(struct(333,3333)));
Time taken: 1.533 seconds
```

```
trino> select * from delta.tiny.test2;
 c1 |  c2  |            c3             
----+------+---------------------------
  2 | NULL | NULL                      
  1 | NULL | NULL                      
  3 | 33   | [{leaf1=333, leaf2=3333}] 
(3 rows)
```
```
spark-sql> show create table test2;
Error in query: Table or permanent view not found: test2; line 1 pos 18;
'ShowCreateTable false, [createtab_stmt#3237]
+- 'UnresolvedTableOrView [test2], SHOW CREATE TABLE, false
```



The existing code seems to have a problem with adding column mapping for arrays of structs.

Now it is 

```
"name": "c3",
            "nullable": true,
            "type": {
                "type": "array",
                "containsNull": true,
                "elementType": {
                    "type": "struct",
                    "fields": [
                        {
                            "metadata": {
                                "delta.columnMapping.id": 5,
                                "delta.columnMapping.physicalName": "col-feb99328-0cb9-4db5-8a92-40eaae0d049f"
                            },
                            "name": "leaf1",
                            "nullable": true,
                            "type": "integer"
                        },
                        {
                            "metadata": {
                                "delta.columnMapping.id": 6,
                                "delta.columnMapping.physicalName": "col-d8f32e3c-d7f2-4270-898b-4827758fa194"
                            },
                            "name": "leaf2",
                            "nullable": true,
                            "type": "integer"
                        }
                    ]
                }
            }
```

instead of 

```
{
            "name": "c3",
            "type": {
                "type": "array",
                "elementType": {
                    "type": "struct",
                    "fields": [
                        {
                            "name": "leaf1",
                            "type": "integer",
                            "nullable": true,
                            "metadata": {
                                "delta.columnMapping.id": 4,
                                "delta.columnMapping.physicalName": "col-00f1d2da-4869-4742-9b62-f1a0fff8ae39"
                            }
                        },
                        {
                            "name": "leaf2",
                            "type": "integer",
                            "nullable": true,
                            "metadata": {
                                "delta.columnMapping.id": 5,
                                "delta.columnMapping.physicalName": "col-12cb0e09-c2e7-4ee0-841e-e6270a56f06f"
                            }
                        }
                    ]
                },
                "containsNull": true
            },
            "nullable": true,
            "metadata": {
                "delta.columnMapping.id": 3,
                "delta.columnMapping.physicalName": "col-269fbe5b-e401-4d5c-8b9c-dfb903e982c0"
            }
        }
```
Notice missing 

```
"metadata": { "delta.columnMapping.id": 3, "delta.columnMapping.physicalName": "col-269fbe5b-e401-4d5c-8b9c-dfb903e982c0" }
```


Repro in spark

```
spark-sql> create table tiny.test3(c1 integer, c2 integer, c3 array<struct<leaf1 integer, leaf2 integer>>) using delta location 's3://trino-ci-test/test3' tblproperties ('delta.columnMapping.mode'='id');
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 5]
```diff
@@ -131,6 +132,7 @@ public static ColumnMappingMode getColumnMappingMode(MetadataEntry metadata)
     public static int getMaxColumnId(MetadataEntry metadata)
     {
         String maxColumnId = metadata.getConfiguration().get(MAX_COLUMN_ID_CONFIGURATION_KEY);
+        requireNonNull(maxColumnId, "maxColumnId is null");
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:135]: Please note that we shouldn't `String.format` in `requireNonNull` method. Also, this is just a defensive check. That's why I chose to report `maxColumnId`.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java (hunk 3 above)
Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:329]: Good catch! I will fix and add more tests.

Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:329]: I reconfirmed the generated transaction log, but it contains the expected id (3) and physicalName. I will add an integration test anyway though.

**Comment by ebyhr** [comment]
Rebased on master to resolve conflicts.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java (hunk 3 above)
Comment (findinpath) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java]: Sorry. I'm reading again the test steps and notice that the problem was in my test scenario :( . I didn't specify the `tiny` schema when trying to see the `SHOW CREATE TABLE` statement output.

Comment (findinpath) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java]: Got a bit confused by the order of the properties - thanks for following up with https://github.com/trinodb/trino/pull/17377

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 6]
```diff
@@ -1212,13 +1221,33 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
...
+                }
+                case ID, NAME -> {
+                    int maxColumnId = getMaxColumnId(handle.getMetadataEntry());
+                    verify(maxColumnId != Integer.MAX_VALUE);
+                    newFieldId.set(maxColumnId + 1);
+                    physicalColumnName = generatePhysicalColumnName();
+                }
+                default -> throw new IllegalArgumentException("Unexpected column mapping mode: " + columnMappingMode);
+            }
+            columnMetadata.put(newColumnMetadata.getName(), metadata.buildOrThrow());
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1242]: `metadata` seems to be unused - it gets initialized with `ImmutableMap.builder()` and then I see only `buildOrThrow()` called on it.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 7]
```diff
@@ -1212,13 +1221,33 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
         try {
             long commitVersion = handle.getReadVersion() + 1;
 
+            ImmutableMap.Builder<String, Map<String, Object>> columnMetadata = ImmutableMap.builder();
+            columnMetadata.putAll(getColumnsMetadata(handle.getMetadataEntry()));
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1225]: Note here that a column may have a primitive type:

```
        {
            "metadata": {
                "delta.columnMapping.id": 1,
                "delta.columnMapping.physicalName": "col-2d56000f-9719-4765-8839-971ceca321c9"
            },
            "name": "col1",
            "nullable": true,
            "type": "integer"
        },
```

or a nested type 

```
        {
            "metadata": {
                "delta.columnMapping.id": 2,
                "delta.columnMapping.physicalName": "col-5f6af4d4-494c-4a36-a638-33b36156f293"
            },
            "name": "c2",
            "nullable": true,
            "type": {
                "type": "struct",
                "fields": [
                    {
                        "metadata": {
                            "delta.columnMapping.id": 3,
                            "delta.columnMapping.physicalName": "col-3347f394-1fcd-4a81-9ebf-95b73a5fe3af"
                        },
                        "name": "leaf1",
                        "nullable": true,
                        "type": "integer"
                    },
                    {
                        "metadata": {
                            "delta.columnMapping.id": 4,
                            "delta.columnMapping.physicalName": "col-eb56f195-96fc-40a8-9799-9178a673b91b"
                        },
                        "name": "leaf2",
                        "nullable": true,
                        "type": "integer"
                    }
                ]
            }
```

Retrieving the only the `metadata` of the existing _base_ element is not enough.


The following scenario corrupts the table:
```
spark-sql> create table tiny.test4 (c1 integer) using delta tblproperties ('delta.columnmapping.mode'='id');
```

```
trino> alter table delta.tiny.test4 add column c2 row(leaf1 integer, leaf2 integer);
ADD COLUMN
trino> alter table delta.tiny.test4 add column c3 row(leaf1 integer, leaf2 integer);
ADD COLUMN
```

Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1225]: Ugh, I was aware of the issue before sending this PR, but forgot to fix it. Thanks for catching that.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 8]
```diff
@@ -391,6 +430,11 @@ private static DeltaLakeColumnMetadata mapColumn(TypeManager typeManager, JsonNo
         return new DeltaLakeColumnMetadata(columnMetadata, fieldId, physicalName, physicalColumnType);
     }
 
+    public static Map<String, Object> getColumnTypes(MetadataEntry metadataEntry)
+    {
+        return getColumnProperties(metadataEntry, node -> node.get("type"));
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:435]:
```suggestion
return getColumnProperties(metadataEntry, node -> OBJECT_MAPPER.convertValue(node.get("type"), new TypeReference<>(){}));
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 9]
```diff
@@ -1218,13 +1246,31 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
...
+                    maxColumnId = new AtomicInteger();
+                    assignIdAndPhysicalName = false;
+                }
+                case ID, NAME -> {
+                    maxColumnId = new AtomicInteger(getMaxColumnId(handle.getMetadataEntry()));
+                    assignIdAndPhysicalName = true;
+                }
+                default -> throw new IllegalArgumentException("Unexpected column mapping mode: " + columnMappingMode);
+            }
+            columnMetadata.put(newColumnMetadata.getName(), ImmutableMap.of());
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1265]: Let's fill the `columnMetadata` here and not in `DeltaLakeSchemaSupport` .
Doing this would spare us from passing `columnMappingMode`, `maxColumnId`, `newColumnName` to `io.trino.plugin.deltalake.transactionlog.DeltaLakeSchemaSupport#serializeSchemaAsJson`

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java [hunk 10]
```diff
@@ -196,6 +296,16 @@ private void testCorruptedTableLocation(String tableName, Path tableLocation, bo
         }
     }
 
+    private MetadataEntry loadMetadataEntry(long entryNumber, Path tableLocation)
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:299]: nit: both private methods can be made `static` - can be done in a follow-up

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 11]
```diff
@@ -1262,11 +1319,15 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
     }
 
     private void appendTableEntries(
+            ColumnMappingMode columnMappingMode,
             long commitVersion,
             TransactionLogWriter transactionLogWriter,
             String tableId,
-            List<DeltaLakeColumnHandle> columns,
+            List<String> columnNames,
+            AtomicInteger newFieldId,
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1327]: `newFieldId` and `newColumnName` are effectively empty for all the calls to this method - let's remove them from here.

**Review by findinpath — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 9 above)
Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1266]: I think it could be doable with a special method in `DeltaLakeSchemaSupport`.

Passing around `maxColumnId` makes it hard to follow when it actually gets modified.

The call to `io.trino.plugin.deltalake.transactionlog.DeltaLakeSchemaSupport#serializeColumnType` increments `maxColumnId` and also `serializeSchemaAsJson` increments it. The order of the calls is also important.

I'd advocate here that the method `serializeSchemaAsJson` does not change the state of `maxColumnId` and it does only what is supposed to do - json serialization and no extra mutation logic.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 12]
```diff
@@ -1219,13 +1239,25 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
         try {
             long commitVersion = handle.getReadVersion() + 1;
 
+            AtomicInteger maxColumnId;
+            switch (columnMappingMode) {
+                case NONE -> {
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1244]: Is the wrapping intentional or rather a left-over?

Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1244]: It's intentional. Otherwise, it throws "InnerAssignment: Inner assignments should be avoided."

**Review by findinpath — APPROVED** [review]
👍  👍

**Comment by ebyhr** [comment]
CI hit concurrent modification error. #17550 will fix it.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSinkProvider.java [hunk 13]
```diff
@@ -99,14 +102,23 @@ public DeltaLakePageSinkProvider(
     public ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorOutputTableHandle outputTableHandle, ConnectorPageSinkId pageSinkId)
     {
         DeltaLakeOutputTableHandle tableHandle = (DeltaLakeOutputTableHandle) outputTableHandle;
+        // TODO support ID and NAME column mapping
+        ColumnMappingMode columnMappingMode = NONE;
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSinkProvider.java:106]: shouldn't this take actual mapping + verify it's NONE?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSinkProvider.java [hunk 14]
```diff
@@ -99,14 +102,23 @@ public DeltaLakePageSinkProvider(
...
+        ImmutableList.Builder<String> columnNames = ImmutableList.builderWithExpectedSize(tableHandle.getInputColumns().size());
+        ImmutableMap.Builder<String, Object> columnTypes = ImmutableMap.builderWithExpectedSize(tableHandle.getInputColumns().size());
+        for (DeltaLakeColumnHandle column : tableHandle.getInputColumns()) {
+            columnNames.add(column.getColumnName());
+            columnTypes.put(column.getColumnName(), serializeColumnType(columnMappingMode, new AtomicInteger(), column.getType()));
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSinkProvider.java:111]: can this be done on the coordinator?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSinkProvider.java (hunk 13 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSinkProvider.java:106]: #17565 will add column mapping to DeltaLakeOutputTableHandle and handle the actual mapping here.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSinkProvider.java (hunk 14 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSinkProvider.java:111]: Moved to DeltaLakeMetadata.

**Comment by ebyhr** [comment]
Addressed comments.

**Review by findepi — APPROVED** [review]
